### ISSUE TITLE
BigGAN: top-k loss weighting (dropping extreme scores)

### ISSUE THREAD

**Issue by gwern** [tensorfork/tensorfork#36]
["Improving GAN Training with Probability Ratio Clipping [PPO] and Sample Reweighting", Wu et al 2020](https://arxiv.org/abs/2006.06900)/["Top-_k_ Training of GANs: Improving GAN Performance by Throwing Away Bad Samples", Sinha et al 2020](https://arxiv.org/abs/2002.06224) introduce a modification to GAN training\*: instead of computing the loss based on the entire minibatch, all generated images good and bad, ignore the worst images. Wu et al 2020 draws on PPO and 'trust regions' to throw away or downweight the worst images, while Sinha et al uses the very crude but effective heuristic of just throwing away the bottom _k_ (gradually increasing to the bottom quarter, in their experiments) of a minibatch.

The logic here is that the worst samples, with the biggest loss, will exert an undue influence on the training of G/D, and destabilize a highly unstable two-player game. (D winning is a perennial issue.) Much like in deep reinforcement learning (note the direct connection between GANs and actor-critic architectures again!), an overreaction to an error can cause a bigger error which then causes even bigger errors, and the system spirals off into chaos and potentially diverges. (It's another RL-ish trick akin to #33, tuning steps by the gradient noise scale.)

Instead, by dropping the wackiest generated points, D doesn't get an easy win but must more closely scrutinize the best samples for errors, leading to stabler training and more refined generation. Sinha reports improvements of ~1 FID point on SAGAN on ImageNet and BigGAN on CIFAR-10. (**Warning**: the gradients should be changed *only* for the Generator; [throwing away D updates does not work at all](https://arxiv.org/pdf/2002.06224.pdf#appendix.C "Appendix C: Applying Top-k Updates to The Critic Hurts Performance"), and at least one re-implementation has failed because they didn't realize it was for G-only, and discovered the hard way that as Sinha et al report, training will fail completely if top-_k_ is applied to D as well.)

The top-k implementation is far easier and simpler than PPO, and PPO is probably overkill, so I prefer top-k. It has already been implemented by lucidrains: https://github.com/lucidrains/stylegan2-pytorch/pull/140 Since it requires no model architecture changes, merely a one or two-line change to the loss computation to drop the lowest-scoring part of the minibatch, Aydao has implemented it themselves & changed over one of their tweaked StyleGAN runs to test it (run `tpu-v3-8-euw4a-201`); as the run was converged, it didn't change much, but the quality did not degrade so it looks like a safe-enough change.

While not a huge win, this should be a very easy win for BigGAN once we have a reliable BigGAN codebase to work with.

\* Another recent sample-reweighting paper: ["MM-nsat: Sample weighting as an explanation for mode collapse in generative adversarial networks", Eide et al 2020](https://arxiv.org/abs/2010.02035).

**Comment by gwern** [comment]
Possibly relevant: ["ODE-GAN: Training Generative Adversarial Networks by Solving Ordinary Differential Equations"](https://arxiv.org/abs/2010.15040#deepmind), Qin et al 2020, shows that using ODEs to step training (instead of regular SGD) makes GAN training *much* more stable and reduces the need for heavy regularization like specnorm. DRL and DL have many connections to ODEs, and the intuition Qin et al offer for why ODEs are useful is that it avoids 'discretization' errors in the training trajectory:

> We find that the main ingredient we need for stable GAN training is to avoid large integration errors, and a simple regulariser on the generator gradients is sufficient to achieve this.This alleviates the need for hard constraints on the functional space of the discriminator (e.g. spectral normalisation [23]) and enables GAN training without advanced optimisation techniques.

The cost is surprisingly low (2x for Runge-Kutta), and the benefit in reducing exploding gradients:

[![xwd-160402063030518](https://user-images.githubusercontent.com/352559/97648851-32e9d300-1a2c-11eb-861f-29c01d08cbd8.png)](https://arxiv.org/pdf/2010.15040.pdf#page=5)


ODE's adaptive error control, top-k, and PPO-style trust regions may all be doing similar things here in terms of stabilizing GAN training.

**Comment by gwern** [comment]
For compare_gan we'd edit somewhere around https://github.com/aydao/compare_gan/blob/ayworking/compare_gan/gans/modular_gan.py#L812

StyleGAN2: https://github.com/lucidrains/stylegan2-pytorch/pull/140/files
